File: tests/algorithms/x/Clojure/machine_learning/automatic_differentiation.clj
(ns main (:refer-clojure :exclude [dual pow_float add sub mul div power main]))

(require 'clojure.set)

(defn in [x coll]
  (cond (string? coll) (clojure.string/includes? coll x) (map? coll) (contains? coll x) (sequential? coll) (some (fn [e] (= e x)) coll) :else false))

(defn padStart [s w p]
  (loop [out (str s)] (if (< (count out) w) (recur (str p out)) out)))

(defn indexOf [s sub]
  (let [idx (clojure.string/index-of s sub)] (if (nil? idx) -1 idx)))

(defn split [s sep]
  (clojure.string/split s (re-pattern sep)))

(defn toi [s]
  (Integer/parseInt (str s)))

(def nowSeed (atom (let [s (System/getenv "MOCHI_NOW_SEED")] (if (and s (not (= s ""))) (Integer/parseInt s) 0))))

(declare dual pow_float add sub mul div power main)

(def ^:dynamic main_a nil)

(def ^:dynamic main_b nil)

(def ^:dynamic main_c nil)

(def ^:dynamic main_d nil)

(def ^:dynamic main_e nil)

(def ^:dynamic main_x nil)

(def ^:dynamic main_y nil)

(def ^:dynamic pow_float_i nil)

(def ^:dynamic pow_float_res nil)

(defn dual [dual_v dual_d]
  (try (throw (ex-info "return" {:v {:deriv dual_d :value dual_v}})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn pow_float [pow_float_base pow_float_exp]
  (binding [pow_float_i nil pow_float_res nil] (try (do (set! pow_float_res 1.0) (set! pow_float_i 0) (while (< pow_float_i pow_float_exp) (do (set! pow_float_res (* pow_float_res pow_float_base)) (set! pow_float_i (+ pow_float_i 1)))) (throw (ex-info "return" {:v pow_float_res}))) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e))))))

(defn add [add_a add_b]
  (try (throw (ex-info "return" {:v {:deriv (+ (:deriv add_a) (:deriv add_b)) :value (+ (:value add_a) (:value add_b))}})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn sub [sub_a sub_b]
  (try (throw (ex-info "return" {:v {:deriv (- (:deriv sub_a) (:deriv sub_b)) :value (- (:value sub_a) (:value sub_b))}})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn mul [mul_a mul_b]
  (try (throw (ex-info "return" {:v {:deriv (+ (* (:deriv mul_a) (:value mul_b)) (* (:deriv mul_b) (:value mul_a))) :value (* (:value mul_a) (:value mul_b))}})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn div [div_a div_b]
  (try (throw (ex-info "return" {:v {:deriv (quot (- (* (:deriv div_a) (:value div_b)) (* (:deriv div_b) (:value div_a))) (* (:value div_b) (:value div_b))) :value (quot (:value div_a) (:value div_b))}})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn power [power_a power_p]
  (try (throw (ex-info "return" {:v {:deriv (* (* (* 1.0 power_p) (pow_float (:value power_a) (- power_p 1))) (:deriv power_a)) :value (pow_float (:value power_a) power_p)}})) (catch clojure.lang.ExceptionInfo e (if (= (ex-message e) "return") (get (ex-data e) :v) (throw e)))))

(defn main []
  (binding [main_a nil main_b nil main_c nil main_d nil main_e nil main_x nil main_y nil] (do (set! main_a (dual 2.0 1.0)) (set! main_b (dual 1.0 0.0)) (set! main_c (add main_a main_b)) (set! main_d (mul main_a main_b)) (set! main_e (div main_c main_d)) (println (str (:deriv main_e))) (set! main_x (dual 2.0 1.0)) (set! main_y (power main_x 3)) (println (str (:deriv main_y))))))

(defn -main []
  (let [rt (Runtime/getRuntime)
    start-mem (- (.totalMemory rt) (.freeMemory rt))
    start (System/nanoTime)]
      (main)
      (System/gc)
      (let [end (System/nanoTime)
        end-mem (- (.totalMemory rt) (.freeMemory rt))
        duration-us (quot (- end start) 1000)
        memory-bytes (Math/abs ^long (- end-mem start-mem))]
        (println (str "{\n  \"duration_us\": " duration-us ",\n  \"memory_bytes\": " memory-bytes ",\n  \"name\": \"main\"\n}"))
      )
    ))

(-main)

File: tests/algorithms/x/Java/machine_learning/automatic_differentiation.java
public class Main {
    static class Dual {
        double value;
        double deriv;
        Dual(double value, double deriv) {
            this.value = value;
            this.deriv = deriv;
        }
        Dual() {}
        @Override public String toString() {
            return String.format("{'value': %s, 'deriv': %s}", String.valueOf(value), String.valueOf(deriv));
        }
    }


    static Dual dual(double v, double d) {
        return new Dual(v, d);
    }

    static double pow_float(double base, long exp) {
        double res = (double)(1.0);
        long i_1 = 0L;
        while ((long)(i_1) < (long)(exp)) {
            res = (double)((double)(res) * (double)(base));
            i_1 = (long)((long)(i_1) + 1L);
        }
        return res;
    }

    static Dual add(Dual a, Dual b) {
        return new Dual((double)(a.value) + (double)(b.value), (double)(a.deriv) + (double)(b.deriv));
    }

    static Dual sub(Dual a, Dual b) {
        return new Dual((double)(a.value) - (double)(b.value), (double)(a.deriv) - (double)(b.deriv));
    }

    static Dual mul(Dual a, Dual b) {
        return new Dual((double)(a.value) * (double)(b.value), (double)((double)(a.deriv) * (double)(b.value)) + (double)((double)(b.deriv) * (double)(a.value)));
    }

    static Dual div(Dual a, Dual b) {
        return new Dual((double)(a.value) / (double)(b.value), (double)(((double)((double)(a.deriv) * (double)(b.value)) - (double)((double)(b.deriv) * (double)(a.value)))) / (double)(((double)(b.value) * (double)(b.value))));
    }

    static Dual power(Dual a, long p) {
        return new Dual(pow_float((double)(a.value), (long)(p)), (double)((double)(((double)(1.0) * (double)(p))) * (double)(pow_float((double)(a.value), (long)((long)(p) - 1L)))) * (double)(a.deriv));
    }

    static void main() {
        Dual a = dual((double)(2.0), (double)(1.0));
        Dual b_1 = dual((double)(1.0), (double)(0.0));
        Dual c_1 = add(a, b_1);
        Dual d_1 = mul(a, b_1);
        Dual e_1 = div(c_1, d_1);
        System.out.println(_p(e_1.deriv));
        Dual x_1 = dual((double)(2.0), (double)(1.0));
        Dual y_1 = power(x_1, 3L);
        System.out.println(_p(y_1.deriv));
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            main();
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        if (v instanceof Double || v instanceof Float) {
            double d = ((Number) v).doubleValue();
            return String.valueOf(d);
        }
        return String.valueOf(v);
    }
}
